File: Java/ByTier/C/2117_C_Cool_Partition.java
// Problem: CF 2117 C - Cool Partition
// https://codeforces.com/contest/2117/problem/C

/*
C. Cool Partition
Algorithms/Techniques: Two-pointer technique with HashSet for tracking element presence
Time Complexity: O(n) per test case, where n is the size of the array
Space Complexity: O(n) for storing elements in hashsets

Problem:
Yousef wants to partition an array into contiguous segments such that each element in a segment also appears in the next segment.
Goal is to maximize the number of such segments.

Approach:
- Use two HashSets: one to track elements in current segment (`set`), another to track all elements seen so far (`supset`)
- Traverse the array and add each element to both sets
- When `set` equals `supset`, we've reached the end of a valid segment. Reset `set` and increment counter.
- Return maximum number of segments found.
*/

import java.util.*;
public class C_Cool_Partition {
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int tcs = sc.nextInt();
        while(tcs-- > 0) {
            int n = sc.nextInt();
            int arr[] = new int[n];
            for(int i=0; i<n; i++) arr[i] = sc.nextInt(); // Read array elements
            int p = 0; // Count of segments
            HashSet<Integer> set = new HashSet<>(); // Elements in current segment
            HashSet<Integer> supset = new HashSet<>(); // All elements seen so far
            for(int i=0; i<n; i++) {
                set.add(arr[i]); // Add to current segment elements
                supset.add(arr[i]); // Add to all seen elements
                if(set.equals(supset)) { // If all elements in current segment also appear in the rest of array so far
                    set.clear(); // Reset current segment
                    p++; // Increment segment count
                }
            }
            System.out.println(p);
        }
        sc.close();
    }
}


// https://github.com/VaHiX/CodeForces/

File: Kotlin/ByRound/2117/2117_C_Cool_Partition.kt
// Problem: CF 2117 C - Cool Partition
// https://codeforces.com/contest/2117/problem/C

/*
C. Cool Partition
Algorithms/Techniques: Two-pointer technique, preprocessing for last occurrence of elements

Time Complexity: O(n) per test case
Space Complexity: O(n) per test case

Purpose: 
This code finds the maximum number of segments that form a "cool partition" of an array.
A cool partition means that every element in a segment must also appear in the next segment.
The approach uses preprocessing to find the last occurrence of each element and then uses
a greedy two-pointer method to determine the maximal number of such segments.
*/

fun solve() {
    val n = readLine()!!.toInt()
    val a = readLine()!!.split(" ").map { it.toInt() }
    val gg = n + 1 // Sentinel value larger than any valid index
    val mo = IntArray(n) { gg } // Stores the last occurrence of each element
    val res = mutableMapOf<Int, Int>() // Maps element to its last seen index
    
    // Preprocessing step: fill 'mo' array with last occurrence indices
    for (i in n - 1 downTo 0) {
        mo[i] = res.getOrDefault(a[i], gg) // Get last occurrence of a[i]
        res[a[i]] = i // Update the most recent index for a[i]
    }
    
    var pos = 0 // Start pointer for current segment
    var ans = 0 // Count of segments
    
    while (pos < n) {
        var end = pos // End pointer of current segment
        while (true) {
            var nn = -1
            // Find the furthest last occurrence among elements in [pos, end]
            for (i in pos..end) {
                if (mo[i] > nn) nn = mo[i]
            }
            ans++ // Increment segment count
            
            if (nn >= n) break // No more elements to extend the segment
            pos = end + 1 // Move start to next segment
            end = nn // Extend end to the furthest last occurrence
        }
        break // Only one partition needed per array (as stated by problem setup)
    }
    
    println(ans)
}

fun main() {
    val t = readLine()!!.toInt()
    repeat(t) { solve() }
}


// https://github.com/VaHiX/CodeForces/